File: grindx/judges/go/two-sum.go
package main

import (
	"encoding/json"
	"math/rand"
	"sort"
)

func main() {
	tc := LoadCases("two-sum")
	basicCases := tc.Cases

	// Generate large cases for TLE detection
	type largeCase struct {
		nums   []int
		target int
	}
	rng := rand.New(rand.NewSource(42))
	var largeCases []largeCase
	for _, n := range []int{10000, 100000} {
		nums := make([]int, n)
		for k := 0; k < n; k++ {
			nums[k] = rng.Intn(2000000001) - 1000000000
		}
		i := rng.Intn(n)
		j := rng.Intn(n)
		for j == i {
			j = (j + 1) % n
		}
		target := nums[i] + nums[j]
		largeCases = append(largeCases, largeCase{nums: nums, target: target})
	}

	total := len(basicCases) + len(largeCases)

	// Run basic cases
	for i, c := range basicCases {
		var nums []int
		json.Unmarshal(c.Input[0], &nums)
		var target int
		json.Unmarshal(c.Input[1], &target)

		numsCopy := make([]int, len(nums))
		copy(numsCopy, nums)
		result := twoSum(numsCopy, target)

		var expected []int
		json.Unmarshal(c.Expected, &expected)

		sortedResult := make([]int, len(result))
		copy(sortedResult, result)
		sort.Ints(sortedResult)

		sortedExpected := make([]int, len(expected))
		copy(sortedExpected, expected)
		sort.Ints(sortedExpected)

		match := len(sortedResult) == len(sortedExpected)
		if match {
			for k := range sortedResult {
				if sortedResult[k] != sortedExpected[k] {
					match = false
					break
				}
			}
		}
		if !match {
			ReportWA(i, c.Input, expected, result, total, c.Category)
		}
	}

	// Run large cases
	for li, lc := range largeCases {
		idx := len(basicCases) + li
		numsCopy := make([]int, len(lc.nums))
		copy(numsCopy, lc.nums)
		result := twoSum(numsCopy, lc.target)

		if len(result) != 2 || result[0] == result[1] ||
			lc.nums[result[0]]+lc.nums[result[1]] != lc.target {
			ReportWA(idx, "large input", lc.target, result, total, "tle")
		}
	}

	ReportAC(total)
}

File: grindx/judges/javascript/two-sum.js
/**
 * Judge for Two Sum — function pattern, unordered comparison.
 */

const path = require("path");
const { loadCases, reportAc, reportWa, loadSolution } = require("./_common");

function generateLargeCases() {
  // Deterministic PRNG (simple LCG, same seed as Python random.seed(42))
  // We use a fixed set of large inputs for TLE detection.
  const cases = [];
  for (const n of [10000, 100000]) {
    const nums = [];
    let seed = 42;
    const rand = () => {
      seed = (seed * 1103515245 + 12345) & 0x7fffffff;
      return (seed % 2000000001) - 1000000000;
    };
    for (let k = 0; k < n; k++) nums.push(rand());
    // Plant a known answer
    const i = Math.abs(rand()) % n;
    let j = Math.abs(rand()) % n;
    while (j === i) j = (j + 1) % n;
    const target = nums[i] + nums[j];
    cases.push({ input: [nums, target], expected_indices: [i, j], category: "tle" });
  }
  return cases;
}

function run(solutionPath) {
  const sol = loadSolution(path.resolve(solutionPath));
  const twoSum = sol.twoSum;
  if (typeof twoSum !== "function") {
    process.stderr.write("Solution does not export a twoSum function\n");
    process.exit(2);
  }

  const tc = loadCases("two-sum");
  const basicCases = tc.cases;
  const largeCases = generateLargeCases();
  const allCases = basicCases.concat(largeCases);
  const total = allCases.length;

  for (let i = 0; i < allCases.length; i++) {
    const c = allCases[i];
    const [nums, target] = c.input;
    const result = twoSum([...nums], target); // copy to prevent mutation

    if (i < basicCases.length) {
      const expected = c.expected;
      if (JSON.stringify([...result].sort((a, b) => a - b)) !==
          JSON.stringify([...expected].sort((a, b) => a - b))) {
        reportWa(i, c, result, total);
      }
    } else {
      // Large cases — validate answer
      if (!Array.isArray(result) || result.length !== 2 ||
          result[0] === result[1] ||
          nums[result[0]] + nums[result[1]] !== target) {
        reportWa(i, c, result, total);
      }
    }
  }
  reportAc(total);
}

run(process.argv[2]);
